File: Algorithm/C++/1000+/1593. Split a String Into the Max Number of Unique Substrings.cc
#include <string>
#include <unordered_set>

class Solution {
public:
    int maxUniqueSplit(std::string s) {
       std::unordered_set<std::string> candidate;
       return dfs(s, candidate, 0);
    }

private:
    int dfs(std::string& s, std::unordered_set<std::string>& candidate, int idx) {
        if (idx == s.length()) {
            return candidate.size();
        }
        int result = 1;
        for (int i = idx; i < s.length(); i++) {
            std::string substr = s.substr(idx, i-idx+1);
            if (candidate.find(substr) == candidate.end()) {
                candidate.insert(substr);
                result = std::max(dfs(s, candidate, i+1), result);
                candidate.erase(substr);
            }
        }
        return result;
    }
};

File: Algorithm/Python/1000+/1593. Split a String Into the Max Number of Unique Substrings.py
class Solution(object):
    def maxUniqueSplit(self, s):
        """
        :type s: str
        :rtype: int
        """
        def dfs(candidate, idx, result):
            if idx == len(s):
                result[0] = max(result[0], len(candidate))
                return
            for i in range(idx, len(s)):
                sub = s[idx: i+1]
                if sub not in candidate:
                    candidate.add(sub)
                    dfs(candidate, i+1, result)
                    candidate.remove(sub)
        
        result, candidate = [1], set()
        dfs(candidate, 0, result)
        return result[0]
        
        # # yet another more concise solution
        # result, candidate = [1], set()
        # def dfs(idx):
        #     if idx == len(s):
        #         result[0] = max(result[0], len(candidate))
        #         return
        #     for i in range(idx, len(s)):
        #         sub = s[idx: i+1]
        #         if sub not in candidate:
        #             candidate.add(sub)
        #             dfs(i+1)
        #             candidate.remove(sub)
        # dfs(0)
        # return result[0]